### PR TITLE
does the mega reaftor yes

### PR DESCRIPTION
**Pull request by Tokyros**
Please accept my mega refactor yes

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -29,66 +36,145 @@ class App extends React.Component<{}, State> {
...
+                dateString: new Date().toString(),
+                userEmail
+            });
+            this.showBanner('Log sent successfuly')
+        } catch(err) {
+            this.showBanner("Failed to send log", 'alert-danger');
+            console.error(err);
         }
-        return <GoogleAuth/>;
     }
```
Comment (Tokyros) [inline src/components/App.tsx]: Hoisted all API calls up to the top level component

File: src/components/App.tsx [hunk 2]
```diff
@@ -29,66 +36,145 @@ class App extends React.Component<{}, State> {
...
+            }
+        });
+
+        setTimeout(() => {
+            const bannerMessageChanged = !this.state.bannerMessage || this.state.bannerMessage.message !== message || this.state.bannerMessage.type !== type;
+            if (!bannerMessageChanged) {
+                this.hideBanner();
+            }
+        }, hideDelay);
+    }
```
Comment (Tokyros) [inline src/components/App.tsx]: Single function to handle showing and hiding of the status bar

File: src/components/App.tsx [hunk 3]
```diff
@@ -29,66 +36,145 @@ class App extends React.Component<{}, State> {
...
-        } else if(!this.currentUser || !validUser(this.currentUser)) {
-            if(!signedIn) {
-                this.setState({ showBanner: true })
-                setTimeout(() => this.setState({ showBanner: false }), 3000);
-            } else {
+    hideBanner = () => {
+        this.setState({
+            bannerMessage: undefined
+        })
+    }
```
Comment (Tokyros) [inline src/components/App.tsx]: Hiding the status banner simply means we do not have a message to show

File: src/components/StatusBanner.tsx [hunk 4]
```diff
@@ -1,31 +1,59 @@
...
+
+    componentWillReceiveProps (nextProps: StatusBannerProps) {
+        if (this.state.visible && !nextProps.bannerMessage) {
+            this.hideBanner();
+        } else if (!this.state.visible && nextProps.bannerMessage) {
+            this.setState({
+                visible: true
+            });
         }
+    }
```
Comment (Tokyros) [inline src/components/StatusBanner.tsx]: This is a bit hacky but the status bar needs to maintain it's properties until it is fully hidden, so we let the component itself handle this change

File: src/components/menu/EnterButton.tsx [hunk 5]
```diff
@@ -1,50 +1,21 @@
...
-                props.trackLogRequest({ success: true, inOffice: true });
-            }
-        } catch(err) {
-            if(err.message === "Network Error") {
-                props.trackLogRequest({ success: true, inOffice: true });
-            } else {
-                props.trackLogRequest({ success: false, inOffice: false, bannerMessage: "Something Went Wrong!" });
-            }
-        }
-    }
```
Comment (Tokyros) [inline src/components/menu/EnterButton.tsx]: These function we're pretty hard to understand, and kind of repeating themselves.
Also this logic really belongs higher up in the chain

File: src/components/App.tsx (hunk 1 above)
Comment (Dor256) [reply src/components/App.tsx]: Ahhh so hoisted is basically lifting state. 
It makes much more sense for the App component to be handling all the api mumbo jumbo, thanks for clearing that up!

Comment (Tokyros) [reply src/components/App.tsx]: image

**Pull request closed without merging**